src/main/resources/db/migration/postgres/V2.4.1.0__data_column_order_states.sql

```diff
@@ -0,0 +1,338 @@
+									  'column conc units', transaction_id, 'codeValue', 'codeValue_column conc units', true, 'ACAS', now(),
+									  0, expt_state_id;
+		-- create hidden value
+		insert into experiment_value (id, code_kind, code_origin, code_type, code_type_and_kind, code_value, deleted, ignored, 
+									  ls_kind, ls_transaction, ls_type, ls_type_and_kind, public_data, recorded_by, recorded_date, 
+									  version, experiment_state_id)
+			select nextval('value_pkseq'), 'boolean', 'ACAS DDICT', 'boolean', 'boolean_boolean', case when hidden then 'TRUE' else 'FALSE' end, false, false,
+									  'hide column', transaction_id, 'codeValue', 'codeValue_hide column', true, 'ACAS', now(),
+									  0, expt_state_id;
+		-- create condition value
+		insert into experiment_value (id, code_kind, code_origin, code_type, code_type_and_kind, code_value, deleted, ignored, 
+									  ls_kind, ls_transaction, ls_type, ls_type_and_kind, public_data, recorded_by, recorded_date, 
+									  version, experiment_state_id)
+			select nextval('value_pkseq'), 'boolean', 'ACAS DDICT', 'boolean', 'boolean_boolean', case when condition then 'TRUE' else 'FALSE' end, false, false,
+									  'condition column', transaction_id, 'codeValue', 'codeValue_condition column', true, 'ACAS', now(),
+									  0, expt_state_id;
+		
+		return expt_state_id;
+	end
+$$ language plpgsql;
+
+-- Create 'data column order' states and values for all experiments which are missing them
+select create_expt_data_column_order_state(experiment_id, (select id from ls_transaction where comments = 'V2.4.1.0 data column order states migration'),
+										   column_order::int, column_type, column_name, units, concentration, conc_units, hidden, false)
+	from tmp_expt_columns where experiment_id in
+		(select e.id from experiment e
+			where not exists (
+				select 1 from experiment_state es where es.ls_kind = 'data column order' and es.experiment_id = e.id
+			)
+		);
+--End Part 2
+
+--Part 3: Create "data column order" states on protocols
+-- This represents a union of all data columns in each of the protocol's experiments
+
+-- Build a temp table of all experiment data column order stuff
+create table tmp_expt_column_order_data as
+   -- Use row_number() to reassign column order at the protocol level, since different experiments may have different columns in different order
+   -- The 'earliest experiment wins', meaning any new endpoints from later experiments will be added later in the column order
+	select *, row_number() over (partition by protocol_id order by experiment_id, column_order) as new_column_order
+	from (
+		-- This query pivots all the 'data column order' experiment values into a table of column_type, column_name, etc.
+		select e.protocol_id,
+			coalesce(col_type.code_value, col_type.string_value) as column_type, --the coalesce is used to account for having the same data in stringValues or in codeValues
```

Comment:
The PR description I think mentions this as:
> Switch several 'stringValue' fields within column order states over to 'codeValue'

But I'm not sure what this is. Is this to fix some local development servers because I don't see where string values exist in this state.

Reply:
Ah, now I see https://github.com/mcneilco/acas/pull/1033

But still curious as to why we added a migration for this for a value which isn't in production. I might still be missing something here.

Reply:
We do need to fix the stringValues on Experiments since they already have these states.
For protocols, yeah, those values are not in production. I added it to cover Hansen's dev environment, and because it was relatively simple to add.

Reply:
I see what you mean now - the coalesce isn't necessary. I'll remove those to simplify it.